reconcile/utils/oc.py

```diff
@@ -62,6 +63,90 @@ class JobNotRunningError(Exception):
     pass
 
 
+class OCDecorators:
+    @classmethod
+    def process_reconcile_time(cls, function):
+        '''
+        Compare current time against bundle commit time and create log
+        and metrics from it.
+
+        This decorator expects an OCProcessReconcileTimeDecoratorMsg
+        object as the only or last element of the decorated function's return
+        value.
+
+        Metrics are generated if the resource doesn't have the
+        qontract.ignore_reconcile_time annotation.
+
+        Log message is created if the following conditions are met:
+
+        * Decorator msg is_log_slow_oc_reconcile is true. This value should
+          come from LOG_SLOW_OC_RECONCILE env variable
+        * Decorator msg slow_oc_reconcile_threshold is less than the time
+          elapsed since the bundle commit timestamp. This value should come
+          from SLOW_OC_RECONCILE_THRESHOLD
+        '''
+
+        @wraps(function)
+        def wrapper(*args, **kwargs):
+            result = function(*args, **kwargs)
+            msg = result[:-1] if isinstance(result, (list, tuple)) else result
+
+            if not isinstance(msg, OCProcessReconcileTimeDecoratorMsg):
+                return result
+
+            running_state = RunningState()
+            commit_time = float(running_state.timestamp)
+            time_spent = time.time() - commit_time
+
+            try:
+                resource_kind = msg.resource['kind']
+                resource_name = msg.resource['metadata']['name']
+                annotations = \
+                    msg.resource['metadata'].get('annotations', {})
+            except KeyError as e:
+                logging.warning(f'Error processing metric: {e}')
+                return result
+
+            function_name = f'{function.__module__}.{function.__qualname__}'
+            ignore_reconcile_time = \
+                annotations.get('qontract.ignore_reconcile_time') == 'true'
+            if not ignore_reconcile_time:
+                reconcile_time.labels(
+                    name=function_name,
+                    integration=running_state.integration
+                ).observe(amount=time_spent)
+
+            if not msg.is_log_slow_oc_reconcile:
+                return
+
+            if time_spent > msg.slow_oc_reconcile_threshold:
+                log_msg = f'Action {function_name} for {resource_kind} ' \
+                          f'{resource_name} in namespace ' \
+                          f'{msg.namespace} from ' \
+                          f'{msg.server} took {time_spent} to ' \
+                          f'reconcile. Commit sha {running_state.commit} ' \
+                          f'and commit ts {running_state.timestamp}.'
+
+                if ignore_reconcile_time:
+                    log_msg += ' Ignored in the metric published.'
+
+                logging.info(log_msg)
+
+            return result
```

Comment:
So, the message object is going through.

Reply:
Yes, we don't really care as it is not being taken into account and it makes things less magical